ta/trusted_keys/entry.c

```diff
@@ -0,0 +1,326 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019-2020, Linaro Limited
+ */
+
+#include <pta_system.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <trusted_keys.h>
+#include <util.h>
+
+#define IV_SIZE			16
+#define TAG_SIZE		16
+#define MAX_BUF_SIZE		512
+
+/*
+ * Acronym:
+ *
+ * TK - Trusted Key
+ */
+
+struct tk_blob_hdr {
+	uint8_t reserved;
+	uint8_t iv[IV_SIZE];
+	uint8_t tag[TAG_SIZE];
+};
+
+static TEE_Result get_random(uint32_t types, TEE_Param params[TEE_NUM_PARAMS])
+{
+	DMSG("Invoked TA_CMD_GET_RANDOM");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!params[0].memref.buffer || !params[0].memref.size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	TEE_GenerateRandom(params[0].memref.buffer, params[0].memref.size);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result derive_unique_key(uint8_t *key, uint16_t key_size,
```

Comment:
It might be useful to have a usage or id parameter too depending on how this function will be used. Compare with `huk_subkey_derive()`.

Reply:
Usage is pretty much fixed here as: `HUK_SUBKEY_UNIQUE_TA` as it relies on underlying system pTA `system_derive_ta_unique_key()` API [here](https://github.com/OP-TEE/optee_os/blob/master/core/pta/system.c#L121).

Reply:
I was thinking for for the sake of this particular TA. In case it ever needs to generate some other unique key.

Reply:
Okay, in that case we can use extra data as this is something that is exposed currently via system pTA. Also, at this point I can't envision the use-case for other unique key so lets rely on this extra buffer instead for the time being.